File: calc.py
#!/usr/bin/env python3
import sys

# TODO: Add support for negative numbers
# binary remainder for division [10101011100011 / 1100101011 = 1101 R: 110110100]
def add(a, b, base):
  base_convert = {16: hex, 8: oct, 2: bin}
  if base in base_convert:
    a, b = int(a, base), int(b, base)
    result = a + b
    return base_convert[base](result)[2:]
  else:
    return "Invalid base"

def subtract(a, b, base):
  base_convert = {16: hex, 8: oct, 2: bin}
  if base in base_convert:
    a, b = int(a, base), int(b, base)
    result = a - b
    return base_convert[base](result)[2:]
  else:
    return "Invalid base"

def multiply(a, b, base):
  base_convert = {16: hex, 8: oct, 2: bin}
  if base in base_convert:
    a, b = int(a, base), int(b, base)
    result = a * b
    return base_convert[base](result)[2:]
  else:
    return "Invalid base"

def divide(a, b, base):
  base_convert = {16: hex, 8: oct, 2: bin}
  if base in base_convert:
    a, b = int(a, base), int(b, base)
    result = a // b
    return base_convert[base](result)[2:]
  else:
    return "Invalid base"


if __name__ == '__main__':
  try:
    arg1, arg2, arg3, arg4 = sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4]
  except:
    print("Usage: python3 calc.py [OPERATOR] [BASE] [DATA1] [DATA2] \ne.g. python3 calc.py add 2 100 110011: add 2 binary numbers 100 and 110011")
    sys.exit(0)

  if arg2 not in ["2", "8", "16"]:
    print('heyy ur almost there! please use an integer(2 for binary, 8 for octal, 16 for hex)')
    sys.exit(0)
  elif arg1 == "add":
    print(add(arg3, arg4, int(arg2)))
  elif arg1 == "subtract":
    print(subtract(arg3, arg4, int(arg2)))
  elif arg1 == "multiply":
    print(multiply(arg3, arg4, int(arg2)))
  elif arg1 == "divide":
    print(divide(arg3, arg4, int(arg2)))
  else:
    print("Invalid operator")

File: calc.go
package main

import (
  "fmt"
  "os"
  "strconv"
)

// TODO: Add support for negative numbers
// binary remainder for division [10101011100011 / 1100101011 = 1101 R: 110110100]
func add(a *string, b *string, base int) *string {
  baseConvert := map[int]func(int64, int) string{16: strconv.FormatInt, 8: strconv.FormatInt, 2: strconv.FormatInt}
  if conv, ok := baseConvert[base]; ok {
    ia, _ := strconv.ParseInt(*a, base, 0)
    ib, _ := strconv.ParseInt(*b, base, 0)
    result := ia + ib
    str := conv(result, base) //str[2:]
    return &str
  } else {
    str := "Invalid base"
    return &str
  }
}

func subtract(a *string, b *string, base int) *string {
  baseConvert := map[int]func(int64, int) string{16: strconv.FormatInt, 8: strconv.FormatInt, 2: strconv.FormatInt}
  if conv, ok := baseConvert[base]; ok {
    ia, _ := strconv.ParseInt(*a, base, 64)
    ib, _ := strconv.ParseInt(*b, base, 64)
    result := ia - ib
    str := conv(result, base)
    return &str
  } else {
    str := "Invalid base"
    return &str
  }
}

func multiply(a *string, b *string, base int) *string {
  baseConvert := map[int]func(int64, int) string{16: strconv.FormatInt, 8: strconv.FormatInt, 2: strconv.FormatInt}
  if conv, ok := baseConvert[base]; ok {
    ia, _ := strconv.ParseInt(*a, base, 64)
    ib, _ := strconv.ParseInt(*b, base, 64)
    result := ia * ib
    str := conv(result, base)
    return &str
  } else {
    str := "Invalid base"
    return &str
  }
}

func divide(a *string, b *string, base int) *string {
  baseConvert := map[int]func(int64, int) string{16: strconv.FormatInt, 8: strconv.FormatInt, 2: strconv.FormatInt}
  if conv, ok := baseConvert[base]; ok {
    ia, _ := strconv.ParseInt(*a, base, 64)
    ib, _ := strconv.ParseInt(*b, base, 64)
    result := ia / ib
    str := conv(result, base)
    return &str
  } else {
    str := "Invalid base"
    return &str
  }
}

func main() {
  if len(os.Args) < 5 {
    fmt.Println("Usage: calc [OPERATOR] [BASE] [DATA1] [DATA2]")
    fmt.Println("e.g. calc add 2 100 110011: add 2 binary numbers 100 and 110011")
    os.Exit(0)
  }

  arg1 := os.Args[1]
  arg2, err := strconv.Atoi(os.Args[2])
  if err != nil {
    fmt.Println("heyy ur almost there! please use an integer(2 for binary, 8 for octal, 16 for hex)")
    os.Exit(0)
  }
  arg3 := os.Args[3]
  arg4 := os.Args[4]

  var result *string

  switch arg1 {
  case "add":
    result = add(&arg3, &arg4, arg2)
  case "subtract":
    result = subtract(&arg3, &arg4, arg2)
  case "multiply":
    result = multiply(&arg3, &arg4, arg2)
  case "divide":
    result = divide(&arg3, &arg4, arg2)
  default:
    fmt.Println("Invalid operator")
    os.Exit(0)
  }
  fmt.Println(*result)
}
